### PR TITLE
Reset adaptive optimizers per batch in ASR attack

### LINKED ISSUE #910: Reset adaptive optimizers in ImperceptibleASRPyTorch attack
The problem does not occur with the default SGD optimizer of the attack. However, when one uses an adaptive optimizer like Adam, the optimizer needs to be resetted after a successful batch of adversarial examples.

In the current case, if the optimization somehow gets stuck in `NaN` values, those values will be propagated to all the following batches. This behavior is not intended. 

**System information (please complete the following information):**
 - Fedora 33
 - Python version 3.6.10
 - ART 1.5.1
 - PyTorch 1.5.1

### PR DESCRIPTION
**Pull request by hesseltuinhof**
# Description

Fixes #910 

## Type of change

Please check all relevant options.

- [x] Bug fix (non-breaking)

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

**Test Configuration**:
- Fedora 33
- Python 3.6.10
- ART 1.5.1
- PyTorch 1.5.1 with DeepSpeech 2.1

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/attacks/evasion/imperceptible_asr/imperceptible_asr_pytorch.py [hunk 1]
```diff
@@ -259,6 +261,16 @@ class only supports targeted attack.
...
 
+            # Next, reset non-SGD optimizers
+            if self.optimizer_1st_stage_arg is not None:
+                self.optimizer_1st_stage = self.optimizer_1st_stage_arg(
+                    params=[self.global_optimal_delta], lr=self.learning_rate_1st_stage
+                )
+            if self.optimizer_2nd_stage_arg is not None:
+                self.optimizer_2nd_stage = self.optimizer_2nd_stage_arg(
+                    params=[self.global_optimal_delta], lr=self.learning_rate_2nd_stage
+                )
```
Comment (beat-buesser) [inline art/attacks/evasion/imperceptible_asr/imperceptible_asr_pytorch.py:272]:
```suggestion
            # Next, reset non-SGD optimizers
            if self._optimizer_1st_stage_arg is not None:
                self.optimizer_1st_stage = self._optimizer_1st_stage_arg(
                    params=[self.global_optimal_delta], lr=self.learning_rate_1st_stage
                )
            if self._optimizer_2nd_stage_arg is not None:
                self.optimizer_2nd_stage = self._optimizer_2nd_stage_arg(
                    params=[self.global_optimal_delta], lr=self.learning_rate_2nd_stage
                )
```

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Hi @hesseltuinhof Thank you very much the updates. I have one proposal, what do you think?

File: art/attacks/evasion/imperceptible_asr/imperceptible_asr_pytorch.py (hunk 1 above)
Comment (hesseltuinhof) [reply art/attacks/evasion/imperceptible_asr/imperceptible_asr_pytorch.py:272]: Great catch Beat! This would have created some troubles otherwise ;)

**Comment by hesseltuinhof** [comment]
Thanks @beat-buesser. That was a really good spot of yours, which slipped in at the end when I made those two attributes private :)

I updated the PR accordingly!

**Comment by beat-buesser** [comment]
Hi @hesseltuinhof Thank you very much, these are great improvements!

The tests TensorFlow 2.4.1v1 and TensorFlow+Lingvo 2.1.0v1 have not run because we changed settings in the repo when the PR was created. I have run them localyl and they are passing.

**Review by beat-buesser — APPROVED** [review]

**Pull request merged**